README.md
# kuaishou-crawler

> As you can see, a crawler for kuaishou pictures and videos

## Latest 

### Version 0.4.1 (2020-03-24)
##### &emsp;View [Change Log](./CHANGELOG.md)

* 现在已经提供exe版本一键执行 [查看](#release) | 或者查看如何运行代码 [查看](#run)
* Python 3.7.3
    * requests
    * json
    * os
    * ~~BeautifulSoup~~
    * re
* 自v0.3.0版本开始，已用面向对象重构，核心代码在`lib/crawler.py`中，启动文件为`crawl.py` / `ks.py`
* 功能：根据用户ID来爬取快手用户的作品，包括视频和图片
    1. 在preset文件(**使用exe版本忽略此文件**)中一行行填写用户id，若缺少文件会自动创建（目前版本已提供自动根据数字id获取真实eid）
        * 分享链接用浏览器打开，根据网址 https://live.kuaishou.com/profile/xxxx 获取id: xxxx~~
        * 或者手机里点开快手用户的头像，底下会告诉你快手号的
    2. 使用时请自己用账号登录快手网站，并使用自己的`cookie['headers']`和`didweb`替换，不保证源代码中对应值可用
        * 关于cookie的did值，你在电脑浏览器中打开快手网站登录后随便打开一个用户的视频作品，然后再地址栏中找到这一项，咱们以giao哥的第一个视频为例
            - https://live.kuaishou.com/u/3x4tn6nm8gnh9xk/3xtd5zf5qbduphc?did=web_12345shiwoluandade
            - 你要保存下来的就是did后面的 `web_12345shiwoluandade`
    3. 因为快手官网会根据cookie，识别你是否在线，爬取的时候要将网页登录并挂着
        * 实测快手网站的用户验证存在30-60分钟左右的有效时长，出现`list index out of range`时极可能是有效期已过，登录网站验证即可
        * 暂且不知道快手官方对过多请求的处理，目前碰到的有上述验证失效，也许也会有请求达到数量会中断请求，此时注释preset中已爬取的用户id，重新开始运行脚本即可
    4. 爬取的视频~~暂时是带水印的(以后考虑获取无水印视频)~~ 是无水印的 **感谢@[tjftjftjf](https://github.com/tjftjftjf)提供手机抓包链接和方法**
* 注意事项：
    * 不考虑提供列表可选的批量下载功能
    * 有需要的合理功能可以issue反馈，看到后会考虑是否修改
    * 如果需要自定义自己的需求，可以拿走代码自行修改，喜欢的话给个star给个follow
    * **本代码仅供学习使用，不可违反法律爬取视频，以及私自盗用搬运视频，后果自负**
    * **本代码仅供学习使用，不可违反法律爬取视频，以及私自盗用搬运视频，后果自负**
    * **本代码仅供学习使用，不可违反法律爬取视频，以及私自盗用搬运视频，后果自负**
    * 重要的说三遍

## Run

0. python3环境与命令行工具
1. 进入项目目录 `cd kuaishou-crawler`
2. 安装依赖 `pip install -r requirements.txt`
3. 运行，有两个版本，`crawl.py`为运行版本，`ks.py`是用于构建exe的版本，当然也可以运行
    * `python crawl.py` / `python ks.py`
    
## Release

https://github.com/oGsLP/kuaishou-crawler/releases

* 下载打包好的exe一键运行(点击download下载即可)
    * ks.exe
    * ks.7z


## Future

* ~~自动根据id获取eid~~ √
* 获取无水印视频 √
* 进一步丰富preset预设文件的可配置选项
* ~~优化代码和log~~ √
* 提供便捷的打包exe √

## Again

**本代码仅供学习使用，不可违反法律爬取视频，以及私自盗用搬运视频，后果自负**

CHANGELOG.md
# Change Log

> 所有版本更新日志会记录在这里

### v0.4.1(2020-03-24)

* 修复用户同一天同名作品无法识别下载的问题
* 修正图片由webp格式转化为jpg的方法

### v0.4.0(2020-03-23)

* 修复id转eid的一些bug
* 从该版本开始，爬取视频均为无水印 **感谢@[tjftjftjf](https://github.com/tjftjftjf)**

### v0.3.0(2020-03-10)

* 修复一些因为用户昵称中存在windows下文件（夹）名非法字符导致os无法写入读取的bug
* 简单看了一点python面向对象，将核心功能提取为crawler类，降低耦合
* 基于crawler类，分出两个文件，一个用于直接在python环境下跑代码，另一个则用于打包好exe一键运行
* 提供exe版本

### v0.2.0(2020-02-26)

* 增加对数字id自动查询转换为eid的支持

### v0.1.1 (2020-02-25)

* 修复`caption`项文字保存为文件名的各种问题
* 增加对`workType` 'ksong'、'single'的支持
* 增加对于爬取过程中请求失败的解决方法
* 增加CHANGELOG.md

### v0.1.0 (2020-02-22)

* 初版

lib/crawler.py
#!usr/bin/env python
# -*- coding:utf-8 -*-
"""
@author: lenovo by XYF
@file: crawler.py
@time: 2020/03/10
@function: 
"""
import requests
import json
import time
import os
import re

INFO = {
    "name": "kuaishou-crawler",
    "author": "oGsLP",
    "repository": "www.github.com/oGsLP/kuaishou-crawler",
    "version": "0.4.1",
    "publishDate": "20-03-24"
}

PROFILE_URL = "https://live.kuaishou.com/profile/"
DATA_URL = "https://live.kuaishou.com/m_graphql"
WORK_URL = "https://v.kuaishou.com/fw/photo/"


class Crawler:
    __param_did = ""

    __headers_web = {
        'accept': '*/*',
        'Accept-Encoding': 'gzip, deflate, br',
        'Accept-Language': 'en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7',
        'Connection': 'keep-alive',
        'Content-Type': 'application/json',
        'Host': 'live.kuaishou.com',
        'Origin': 'https://live.kuaishou.com',
        'Sec-Fetch-Mode': 'cors',
        'Sec-Fetch-Site': 'same-origin',
        # User-Agent/Cookie 根据自己的电脑修改
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.130 Safari/537.36',
        'Cookie': ''
    }
    __headers_mobile = {
        'User-Agent': 'Mozilla/5.0 (Linux; U; Android 8.1.0; zh-cn; BLA-AL00 Build/HUAWEIBLA-AL00) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/57.0.2987.132 MQQBrowser/8.9 Mobile Safari/537.36'
    }

    __crawl_list = []

    __date_cache = ""
    __date_pic_count = 0

    def __init__(self, prod=True):
        self.__intro()
        if prod:
            time.sleep(5)
        else:
            self.__read_preset()

    def set_did(self, did):
        self.__param_did = did
        self.__headers_web['Cookie'] = 'did=' + did + "; userId="

    def crawl(self):
        print("准备开始爬取，共有%d个用户..." % len(self.__crawl_list))
        print()
        time.sleep(2)
        for uid in self.__crawl_list:
            self.__date_cache = ""
            self.__date_count = 0
            self.__crawl_user(uid)

    def add_to_list(self, uid):
        self.__crawl_list.append(uid)

    def __crawl_user(self, uid):
        if uid.isdigit():
            uid = self.__switch_id(uid)

        payload = {"operationName": "privateFeedsQuery",
                   "variables": {"principalId": uid, "pcursor": "", "count": 999},
                   "query": "query privateFeedsQuery($principalId: String, $pcursor: String, $count: Int) {\n  privateFeeds(principalId: $principalId, pcursor: $pcursor, count: $count) {\n    pcursor\n    list {\n      id\n      thumbnailUrl\n      poster\n      workType\n      type\n      useVideoPlayer\n      imgUrls\n      imgSizes\n      magicFace\n      musicName\n      caption\n      location\n      liked\n      onlyFollowerCanComment\n      relativeHeight\n      timestamp\n      width\n      height\n      counts {\n        displayView\n        displayLike\n        displayComment\n        __typename\n      }\n      user {\n        id\n        eid\n        name\n        avatar\n        __typename\n      }\n      expTag\n      __typename\n    }\n    __typename\n  }\n}\n"}
        res = requests.post(DATA_URL, headers=self.__headers_web, json=payload)

        works = json.loads(res.content.decode(encoding='utf-8', errors='strict'))['data']['privateFeeds']['list']

        if not os.path.exists("../data"):
            os.makedirs("../data")

        # 这两行代码将response写入json供分析
        # with open("data/" + uid + ".json", "w") as fp:
        #     fp.write(json.dumps(works, indent=2))

        # 防止该用户在直播，第一个作品默认为直播，导致获取信息为NoneType
        if works[0]['id'] is None:
            works.pop(0)
        name = re.sub(r'[\\/:*?"<>|\r\n]+', "", works[0]['user']['name'])

        dir = "data/" + name + "(" + uid + ")/"
        # print(len(works))
        if not os.path.exists(dir):
            os.makedirs(dir)

        # if not os.path.exists(dir + ".list"):
        #     print("")

        print("开始爬取用户 " + name + "，保存在目录 " + dir)
        print(" 共有" + str(len(works)) + "个作品")

        for j in range(len(works)):
            self.__crawl_work(uid, dir, works[j], j + 1)
            time.sleep(1)

        print("用户 " + name + "爬取完成!")
        print()
        time.sleep(1)

    '''
    快手分为五种类型的作品，在作品里面表现为workType属性
     * 其中两种图集: `vertical`和`multiple`，意味着拼接长图和多图，所有图片的链接在imgUrls里
     * 一种单张图片: `single` 图片链接也在imgUrls里
     * K歌: `ksong` 图片链接一样，不考虑爬取音频...
     * 视频: `video` 需要解析html获得视频链接
    '''

    def __crawl_work(self, uid, dir, work, wdx):
        w_type = work['workType']
        w_caption = re.sub(r"\s+", " ", work['caption'])
        w_name = re.sub(r'[\\/:*?"<>|\r\n]+', "", w_caption)[0:24]
        w_time = time.strftime('%Y-%m-%d', time.localtime(work['timestamp'] / 1000))
        w_index = ""
        if self.__date_cache == w_time:
            self.__date_count = self.__date_count + 1
            if self.__date_count > 0:
                w_index = "(%d)" % self.__date_count
        else:
            self.__date_cache = w_time
            self.__date_count = 0

        if w_type == 'vertical' or w_type == 'multiple' or w_type == "single" or w_type == 'ksong':
            w_urls = work['imgUrls']
            l = len(w_urls)
            print("  " + str(wdx) + ")图集作品：" + w_caption + "，" + "共有" + str(l) + "张图片")
            for i in range(l):
                p_name = w_time + w_index + "_" + w_name + "_" + str(i + 1)+'.jpg'
                pic = dir + p_name
                if not os.path.exists(pic):
                    r = requests.get(w_urls[i].replace("webp","jpg"))
                    r.raise_for_status()
                    with open(pic, "wb") as f:
                        f.write(r.content)
                    print("    " + str(i + 1) + "/" + str(l) + " 图片 " + p_name + " 下载成功 √")
                else:
                    print("    " + str(i + 1) + "/" + str(l) + " 图片 " + p_name + " 已存在 √")
        elif w_type == 'video':
            w_url = WORK_URL + work['id']
            res = requests.get(w_url, headers=self.__headers_mobile,
                               params={"fid": 1841409882, "cc": "share_copylink", "shareId": "143108986354"})
            html = res.text
            waitreplace = work['id'] + '&#34.*?&#34;srcNoMark&#34;:&#34;(.*?)&#34;'

            v_url = re.findall(waitreplace, html)
            # pattern = re.compile(r"playUrl", re.MULTILINE | re.DOTALL)
            # script = soup.find("script", text=pattern)
            # s = pattern.search(script.text).string
            # v_url = s.split('playUrl":"')[1].split('.mp4')[0].encode('utf-8').decode('unicode-escape') + '.mp4'
            try:
                print("  " + str(wdx) + ")视频作品：" + w_caption)
            except:
                print("  这里似乎有点小错误，已跳过")
            v_name = w_time + w_index + "_" + w_name + ".mp4"
            video = dir + v_name

            if v_url:
                if not os.path.exists(video):
                    r = requests.get(v_url[0])
                    r.raise_for_status()

                    with open(video, "wb") as f:
                        f.write(r.content)
                    print("    视频 " + v_name + " 下载成功 √")
                else:
                    print("    视频 " + v_name + " 已存在 √")
            else:
                print("未找到视频")
        else:
            print("错误的类型")

    def __read_preset(self):
        p_path = "preset"
        if not os.path.exists(p_path):
            print("创建预设文件 preset ...")
            open(p_path, "w")
        if not os.path.getsize(p_path):
            print("请在预设文件 preset 中记录需要爬取的用户id，一行一个")
            exit(0)
        with open(p_path, "r") as f:
            for line in f:
                if line[0] != "#":
                    self.__crawl_list.append(line.strip())

    def __switch_id(self, uid):
        payload = {"operationName": "SearchOverviewQuery",
                   "variables": {"keyword": uid, "ussid": None},
                   "query": "query SearchOverviewQuery($keyword: String, $ussid: String) {\n  pcSearchOverview(keyword: $keyword, ussid: $ussid) {\n    list {\n      ... on SearchCategoryList {\n        type\n        list {\n          categoryId\n          categoryAbbr\n          title\n          src\n          __typename\n        }\n        __typename\n      }\n      ... on SearchUserList {\n        type\n        ussid\n        list {\n          id\n          name\n          living\n          avatar\n          sex\n          description\n          counts {\n            fan\n            follow\n            photo\n            __typename\n          }\n          __typename\n        }\n        __typename\n      }\n      ... on SearchLivestreamList {\n        type\n        lssid\n        list {\n          user {\n            id\n            avatar\n            name\n            __typename\n          }\n          poster\n          coverUrl\n          caption\n          id\n          playUrls {\n            quality\n            url\n            __typename\n          }\n          quality\n          gameInfo {\n            category\n            name\n            pubgSurvival\n            type\n            kingHero\n            __typename\n          }\n          hasRedPack\n          liveGuess\n          expTag\n          __typename\n        }\n        __typename\n      }\n      __typename\n    }\n    __typename\n  }\n}\n"}

        res = requests.post(DATA_URL, headers=self.__headers_web, json=payload)
        dt = json.loads(res.content.decode(encoding='utf-8', errors='strict'))['data']
        # with open("data/jj_" + uid + ".json", "w") as fp:
        #     fp.write(json.dumps(dt, indent=2))

        return dt['pcSearchOverview']['list'][1]['list'][0]['id']

    def __intro(self):
        print()
        print("|  %s (v%s %s)" % (INFO["name"], INFO["version"], INFO["publishDate"]))
        print("|  本程序由%s提供, %s, 喜欢的话可以给个star >_<" % (INFO["author"], INFO["repository"]))
        print()
